typhos/tools/console.py

```diff
@@ -61,62 +74,51 @@ def sizeHint(self):
 
     def shutdown(self):
         """Shutdown the Jupyter Kernel"""
-        client = self.kernel.kernel_client
-        if client.channels_running:
-            logger.debug("Stopping Jupyter Client")
-            # Stop channels in the background
-            t = threading.Thread(target=client.stop_channels)
-            t.start()
-            self.kernel.kernel_manager.shutdown_kernel()
-        else:
-            logger.debug("Kernel is already shutdown.")
-
-
-try:
-    import happi
-
-    def add_device(obj, device):
-        # Needs metadata
-        if not hasattr(device, 'md'):
-            logger.error("Device %r has no stored metadata. "
-                         "Unable to load in TyphosConsole",
-                         device)
+        client = self.jupyter_widget.kernel_client
+        if self._shutting_down:
+            logger.debug("Kernel is already shutting down")
             return
-        # Create a temporary file
-        name = hashlib.md5(str(localtime()).encode('utf-8')).hexdigest()
-        name = os.path.join(tempfile.gettempdir(), name)
+
+        self._shutting_down = True
+        logger.debug("Stopping Jupyter Client")
+
+        def cleanup():
+            self.jupyter_widget.kernel_manager.shutdown_kernel()
+            self.kernel_shut_down.emit()
+
+        client.stop_channels()
+        QtCore.QTimer.singleShot(0, cleanup)
+
+    def add_device(self, device):
+        # Add the device after a short delay to allow the console widget time
+        # to get initialized
+        QtCore.QTimer.singleShot(
+            100, lambda device=device: self._add_device(device)
+        )
```

Comment:
Timers with magic numbers always make me nervous

Reply:
Not sure what we could do better. Perhaps watch the IPython console output until we see a prompt `In [1]:` and emit a signal?

I dug a bit, but didn't see much else in the interface to aid in synchronization. Rather, programmatic interaction with the kernel in general seemed to be somewhat lacking.

Reply:
If there's nothing better, there's nothing better. 🤷

Reply:
Let me give the above a try and see how it works...